google-cloud-spanner-cassandra/src/main/java/com/google/cloud/spanner/adapter/DriverConnectionHandler.java

```diff
@@ -270,4 +282,29 @@ private Optional<byte[]> prepareAttachmentForQueryId(
   private static String constructKey(byte[] queryId) {
     return PREPARED_QUERY_ID_ATTACHMENT_PREFIX + new String(queryId, StandardCharsets.UTF_8);
   }
+
+  private static ApiCallContext createContext(Frame frame) {
+    ApiCallContext context = GrpcCallContext.createDefault();
+    if (isDML(frame)) {
+      context = context.withExtraHeaders(ROUTE_TO_LEADER_HEADER_MAP);
+    }
+    return context;
+  }
+
+  /** Whether the statement decoded in the given Frame is a DML statement. */
+  private static boolean isDML(Frame frame) {
```

Comment:
It would be great if we can combine this method with `tryPrepareCassandraAttachments`, we can avoid redundant frame message conversion.

Reply:
done